### DISCUSSION TITLE
YOLOv5 Focus() Layer

### DISCUSSION THREAD

**Discussion by glenn-jocher** [ultralytics/yolov5#3181] [Ideas]
I've received lots of interest in the **YOLOv5 🚀 Focus layer**, so I've written a short documentation here on it. I created the Focus layer myself when evolving the YOLOv3 architecture into YOLOv5, and did not adopt it from other sources. The **main purpose** of the Focus layer is to **reduce layers, reduce parameters, reduce FLOPS, reduce CUDA memory, increase forward and backward speed while minimally impacting mAP**.
https://github.com/ultralytics/yolov5/blob/17b0f71538d3e9990e0e6c4b5c7c48375956efa3/models/common.py#L163-L171

The YOLOv5 Focus layer replaces the **first 3 YOLOv3 layers** with a **single layer**:
<img width="1033" alt="Focus Layer" src="https://user-images.githubusercontent.com/26833433/118395682-f38f7680-b64b-11eb-8fe4-6d6103ba8176.png">

I settled on the current Focus layer design after a significant effort profiling alternative designs to the YOLOv3 input layers, both for immediate forward/backward/memory profiling results, and also comparing full 300 epoch COCO trainings to determine the effect on mAP. You can profile the Focus layer very easily against the YOLOv3 original layers replaced using the [YOLOv5 `profile()` function](https://github.com/ultralytics/yolov5/blob/17b0f71538d3e9990e0e6c4b5c7c48375956efa3/utils/torch_utils.py#L96-L102):

```python
# Profile
import torch.nn as nn
from models.common import Focus, Conv, Bottleneck
from utils.torch_utils import profile 

m1 = Focus(3, 64, 3)  # YOLOv5 Focus layer
m2 = nn.Sequential(Conv(3, 32, 3, 1), Conv(32, 64, 3, 2), Bottleneck(64, 64))  # YOLOv3 first 3 layers

results = profile(input=torch.randn(16, 3, 640, 640), ops=[m1, m2], n=10, device=0)  # profile both 10 times at batch-size 16
```

In our [YOLOv5 Google Colab notebook](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) I get the following result:

GPU `device=0`
```python
YOLOv5 🚀 v5.0-405-gfad57c2 torch 1.9.0+cu102 CUDA:0 (Tesla T4, 15109.75MB)

      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
        7040       23.07         2.259         16.65          54.1       (16, 3, 640, 640)      (16, 64, 320, 320)  # Focus
       40160       140.7         7.522         77.86         331.9       (16, 3, 640, 640)      (16, 64, 320, 320)
```

CPU `device='cpu'`
```python
      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
        7040       23.07         0.000         882.1          2029       (16, 3, 640, 640)      (16, 64, 320, 320)  # Focus
       40160       140.7         0.000          4513          8565       (16, 3, 640, 640)      (16, 64, 320, 320)
```

**Comment by developer0hye** [comment]
@glenn-jocher 

Is there comparative results between the YOLOV5 with YOLOv3 original layers and Focus layer in terms of mAP?

**Reply by glenn-jocher** [reply]
@developer0hye yes we have lots of previous studies, but these are all outdated now due to the accumulated design changes in the last year. I'll go ahead and run a new training to compare results on two fully trained YOLOv5m6 models with and without the Focus layer.

[YOLOv5m6](https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5m6.yaml):
`Model Summary: 503 layers, 35917020 parameters, 35917020 gradients, 52.5 GFLOPS`

[YOLOv5m6_noFocus](https://github.com/ultralytics/yolov5/commit/c359b64e61f5ceccca5ae617d17eb2c0d51dae42):
`Model Summary: 515 layers, 35934564 parameters, 35934564 gradients, 56.5 GFLOPS`

**Reply by developer0hye** [reply]
@glenn-jocher Thanks a lot!

**Reply by developer0hye** [reply]
@glenn-jocher 
Is there any progressive on this issue?

**Reply by glenn-jocher** [reply]
@developer0hye no, sorry, I started one training which crashed for some reason and I didn't have time to investigate.

**Reply by developer0hye** [reply]
@glenn-jocher It's okay! Thanks!

**Reply by luohao123** [reply]
@glenn-jocher Does it have a result? what's the gap between w/o Focus

**Comment by hussein-MA** [comment]
@glenn-jocher Thank you for your great work
I have question, why I do not see the calling for focus in yolov5s.yaml?
backbone:
[from, number, module, args]

[[-1, 1, Conv, [64, 6, 2, 2]], # 0-P1/2
[-1, 1, Conv, [128, 3, 2]], # 1-P2/4
[-1, 3, C3, [128]],
[-1, 1, Conv, [256, 3, 2]], # 3-P3/8
[-1, 6, C3, [256]],
[-1, 1, Conv, [512, 3, 2]], # 5-P4/16
[-1, 9, C3, [512]],
[-1, 1, Conv, [1024, 3, 2]], # 7-P5/32
[-1, 3, C3, [1024]],
[-1, 1, SPPF, [1024, 5]], # 9
]

Is the first line [[-1, 1, Conv, [64, 6, 2, 2]], # 0-P1/2 represents the Focus?
May I know where the focus is called?

**Reply by glenn-jocher** [reply]
@hussein-MA Focus() was replaced by Conv() for better exportability in release v6.0. See release notes for details:
https://github.com/ultralytics/yolov5/releases/tag/v6.0

**Reply by hussein-MA** [reply]
@glenn-jocher Thank you

**Comment by soumya997** [comment]
What is the purpose of using a focus structure? slicing the image into 4 parts and then passing them through a conv layer, how does it help, in yolov5? thanks.

cc: @glenn-jocher

**Comment by bechir23** [comment]
when i added focus layer in yolov10 i don t get the gflops  anymore i get only the number of parameters. Does an
![Screenshot 2024-11-13 211734](https://github.com/user-attachments/assets/3ae934e6-d4e4-4528-a652-9eacd8b42383)
